geode-core/src/main/java/org/apache/geode/internal/net/NioSslEngine.java

```diff
@@ -368,17 +365,15 @@ public void doneReadingDirectAck(ByteBuffer unwrappedBuffer) {
     // read-operations
   }
 
-  @Override
-  public synchronized boolean isClosed() {
-    return closed;
-  }
-
   @Override
   public synchronized void close(SocketChannel socketChannel) {
```

Comment:
Will this "synchronized" sync with wrap() method's shareOutputBuffer?

wrap() and close() have to sync with each other, otherwise it will encounter bug Gem-3116.

Reply:
The `close()` method is a critical section. And after a thread has exited that critical section, the next thread entering it will skip all the actual close logic.

There is no possible interference between the close logic and e.g. `wrap()`, specifically in their use of the output buffer a.k.a. `myNetData` (the `ByteBuffer` managed by `outputSharing`) because both `wrap()` and `close()` acquire a reference to that buffer via `ByteBufferSharingImpl.open()` and `ByteBufferSharingImpl.open(long time, TimeUnit unit)` respectively. Those methods call `lock()`/`tryLock(long time, TimeUnit unit)` on the `ReentrantLock` before returning.

Does that help?